src/finiteVolume/processes/soot.cpp

```diff
@@ -0,0 +1,397 @@
+    pointInformation.speciesSensibleEnthalpyFunction = eos->GetThermodynamicTemperatureMassFractionFunction(eos::ThermodynamicProperty::SpeciesSensibleEnthalpy, mockFields);
+}
+
+PetscErrorCode ablate::finiteVolume::processes::Soot::ComputeSootChemistryPreStep(ablate::finiteVolume::FiniteVolumeSolver& flow, TS flowTs, PetscReal time, bool initialStage, Vec locX, void* ctx) {
+    PetscFunctionBeginUser;
+    if (!initialStage) {
+        PetscFunctionReturn(0);
+    }
+    auto soot = (ablate::finiteVolume::processes::Soot*)ctx;
+
+    // Get the valid cell range over this region
+    domain::Range cellRange;
+    flow.GetCellRange(cellRange);
+
+    // get the dim
+    PetscInt dim;
+    PetscCall(DMGetDimension(flow.GetSubDomain().GetDM(), &dim));
+
+    // store the current dt
+    PetscReal dt;
+    PetscCall(TSGetTimeStep(flowTs, &dt));
+
+    // get access to the underlying data for the flow
+    const auto& flowEulerId = flow.GetSubDomain().GetField(ablate::finiteVolume::CompressibleFlowFields::EULER_FIELD);
+    const auto& flowDensityYiId = flow.GetSubDomain().GetField(ablate::finiteVolume::CompressibleFlowFields::DENSITY_YI_FIELD);
+    const auto& flowDensityProgressId = flow.GetSubDomain().GetField(ablate::finiteVolume::CompressibleFlowFields::DENSITY_PROGRESS_FIELD);
+
+    // get the flowSolution from the ts
+    const PetscScalar* solutionArray;
+    PetscCall(VecGetArrayRead(locX, &solutionArray));
+
+    // get the aux array for temperature
+    const auto& temperatureField = flow.GetSubDomain().GetField(ablate::finiteVolume::CompressibleFlowFields::TEMPERATURE_FIELD);
+    auto temperatureDm = flow.GetSubDomain().GetFieldDM(temperatureField);
+    auto temperatureVec = flow.GetSubDomain().GetVec(temperatureField);
+    const PetscScalar* temperatureArray;
+    PetscCall(VecGetArrayRead(temperatureVec, &temperatureArray));
+
+    // Get access to the chemistry source.  This is sized for euler + nspec
+    PetscScalar* sourceArray;
+    PetscCall(VecGetArray(soot->sourceVec, &sourceArray));
+
+    // get an easy reference to the point information
+    auto& pointInformation = soot->pointInformation;
+
+    // March over each cell
+    for (PetscInt c = cellRange.start; c < cellRange.end; ++c) {
+        // if there is a cell array, use it, otherwise it is just c
+        const PetscInt cell = cellRange.GetPoint(c);
+
+        // Get the current state variables for this cell
+        const PetscScalar* conserved = nullptr;
+        PetscCall(DMPlexPointGlobalRead(flow.GetSubDomain().GetDM(), cell, solutionArray, &conserved));
+
+        // If a real cell (not ghost)
+        if (conserved) {
+            // store the data for the chemistry ts (T, Yi...)
+            PetscReal* temperature;
+            PetscCall(DMPlexPointLocalFieldRead(temperatureDm, cell, temperatureField.id, temperatureArray, &temperature));
+
+            if (*temperature > soot->thresholdTemperature) {
```

Comment:
If we're doing things like this is it worth considering dynamic load balancing of the mesh? I wonder how hard that would be to implement.